packages/grid/x-data-grid/src/hooks/features/density/useGridDensity.tsx

```diff
@@ -18,51 +21,117 @@ const getUpdatedDensityState = (
   newDensity: GridDensity,
   newHeaderHeight: number,
   newRowHeight: number,
+  newMaxDepth: number,
+  newHeaderGroupingRowHeight: number,
 ): GridDensityState => {
   switch (newDensity) {
     case GridDensityTypes.Compact:
       return {
         value: newDensity,
         headerHeight: Math.floor(newHeaderHeight * COMPACT_DENSITY_FACTOR),
         rowHeight: Math.floor(newRowHeight * COMPACT_DENSITY_FACTOR),
+        headerGroupingRowHeight: Math.floor(newHeaderGroupingRowHeight * COMPACT_DENSITY_FACTOR),
+        headerGroupingMaxDepth: newMaxDepth,
         factor: COMPACT_DENSITY_FACTOR,
       };
     case GridDensityTypes.Comfortable:
       return {
         value: newDensity,
         headerHeight: Math.floor(newHeaderHeight * COMFORTABLE_DENSITY_FACTOR),
         rowHeight: Math.floor(newRowHeight * COMFORTABLE_DENSITY_FACTOR),
+        headerGroupingRowHeight: Math.floor(
+          newHeaderGroupingRowHeight * COMFORTABLE_DENSITY_FACTOR,
+        ),
+        headerGroupingMaxDepth: newMaxDepth,
         factor: COMFORTABLE_DENSITY_FACTOR,
       };
     default:
       return {
         value: newDensity,
         headerHeight: newHeaderHeight,
         rowHeight: newRowHeight,
+        headerGroupingRowHeight: newHeaderGroupingRowHeight,
+        headerGroupingMaxDepth: newMaxDepth,
         factor: 1,
       };
   }
 };
 
 export const densityStateInitializer: GridStateInitializer<
-  Pick<DataGridProcessedProps, 'density' | 'headerHeight' | 'rowHeight'>
-> = (state, props) => ({
-  ...state,
-  density: getUpdatedDensityState(props.density, props.headerHeight, props.rowHeight),
-});
+  Pick<
+    DataGridProcessedProps,
+    'density' | 'headerHeight' | 'rowHeight' | 'headerGroupingRowHeight' | 'columnGroupingModel'
+  >
+> = (state, props) => {
+  // TODO: think about improving this initialization. Could it be done in the useColumn initializer?
+  // TODO: ma,age to remove ts-ignore
+  let maxDepth: number;
+  if (props.columnGroupingModel == null || Object.keys(props.columnGroupingModel).length === 0) {
+    maxDepth = 0;
+  } else {
+    const unwrappedGroupingColumnModel = unwrapGroupingColumnModel(props.columnGroupingModel);
+
+    // @ts-ignore
+    const visibleColumns = state.columns.all.filter(
```

Comment:
You can remove the ts-ignore by just forcing the non-nullability of the variables

```tsx
    const visibleColumns = state.columns!.all!.filter(
      (field) => state.columns!.columnVisibilityModel![field!] !== false,
    );
```

By the way, the `field!` should not be required, the `DeePartial` interfact should not make array nullable, just objects.
I'll open a PR